File: typescript/src/context/surgeon.ts
import { ContextBlock, ContextWeight } from '../types.js';

export enum CompressionStrategy {
  SlidingWindow = 'sliding_window',
  WeightedPrune = 'weighted_prune',
  Hybrid = 'hybrid',
}

export interface SurgeonConfig {
  strategy: CompressionStrategy;
  targetTokens: number;
  preserveRecent: number;
  keepRecent?: number;
}

export interface SurgeonResult {
  blocks: ContextBlock[];
  tokensBefore: number;
  tokensAfter: number;
  tokensSaved: number;
  strategyApplied: string;
}

export function compress(blocks: ContextBlock[], config: SurgeonConfig): SurgeonResult {
  const tokensBefore = blocks.reduce((s, b) => s + b.tokenCount, 0);
  let kept: ContextBlock[];

  if (config.strategy === CompressionStrategy.SlidingWindow) {
    kept = slidingWindow(blocks, config.keepRecent ?? config.preserveRecent, config.preserveRecent);
  } else if (config.strategy === CompressionStrategy.WeightedPrune) {
    kept = weightedPrune(blocks, config.targetTokens, config.preserveRecent);
  } else {
    const afterPrune = weightedPrune(blocks, config.targetTokens, config.preserveRecent);
    const total = afterPrune.reduce((s, b) => s + b.tokenCount, 0);
    kept = total > config.targetTokens
      ? slidingWindow(afterPrune, config.keepRecent ?? config.preserveRecent, config.preserveRecent)
      : afterPrune;
  }

  const tokensAfter = kept.reduce((s, b) => s + b.tokenCount, 0);
  return {
    blocks: kept,
    tokensBefore,
    tokensAfter,
    tokensSaved: tokensBefore - tokensAfter,
    strategyApplied: config.strategy,
  };
}

function slidingWindow(blocks: ContextBlock[], keepRecent: number, preserveRecent: number): ContextBlock[] {
  if (keepRecent === 0) return [...blocks];
  const alwaysKeep = Math.max(preserveRecent, keepRecent);
  if (blocks.length <= alwaysKeep) return [...blocks];

  const pinned = blocks.filter(b => b.weight >= ContextWeight.High);
  const pinnedIds = new Set(pinned.map(b => b.id));
  const recentStart = Math.max(0, blocks.length - keepRecent);
  const result = [...pinned];
  for (const b of blocks.slice(recentStart)) {
    if (!pinnedIds.has(b.id)) result.push(b);
  }
  return result;
}

function weightedPrune(blocks: ContextBlock[], targetTokens: number, preserveRecent: number): ContextBlock[] {
  const total = blocks.reduce((s, b) => s + b.tokenCount, 0);
  if (total <= targetTokens) return [...blocks];

  const recentSlice = preserveRecent > 0 ? blocks.slice(-preserveRecent) : [];
  const preserveIds = new Set(recentSlice.map(b => b.id));
  const removable = blocks
    .filter(b => !preserveIds.has(b.id) && b.weight < ContextWeight.High)
    .sort((a, b) => a.weight !== b.weight ? a.weight - b.weight : a.createdAt - b.createdAt);

  const toRemove = new Set<string>();
  let running = total;
  for (const b of removable) {
    if (running <= targetTokens) break;
    running -= b.tokenCount;
    toRemove.add(b.id);
  }
  return blocks.filter(b => !toRemove.has(b.id));
}

export function scoreRelevance(block: ContextBlock, task: string): number {
  const taskWords = new Set(task.split(/\s+/).filter(Boolean));
  if (taskWords.size === 0) return 0;
  const matches = block.content.split(/\s+/).filter(w => taskWords.has(w)).length;
  return Math.min(matches / taskWords.size, 1.0);
}

export function detectContradiction(a: ContextBlock, b: ContextBlock): boolean {
  const aLow = a.content.toLowerCase();
  const bLow = b.content.toLowerCase();
  for (const word of aLow.split(/\s+/)) {
    if (word.length > 4 && bLow.includes(`not ${word}`)) return true;
  }
  return false;
}

File: python/src/truss/context/surgeon.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum

from truss.types import ContextBlock, ContextWeight


class CompressionStrategy(Enum):
    SLIDING_WINDOW = "sliding_window"
    WEIGHTED_PRUNE = "weighted_prune"
    HYBRID = "hybrid"


@dataclass
class SurgeonConfig:
    strategy: CompressionStrategy = CompressionStrategy.HYBRID
    target_tokens: int = 8_000
    preserve_recent: int = 5
    keep_recent: int = 0


@dataclass
class SurgeonResult:
    blocks: list[ContextBlock]
    tokens_before: int
    tokens_after: int
    tokens_saved: int
    strategy_applied: str


def compress(blocks: list[ContextBlock], config: SurgeonConfig) -> SurgeonResult:
    tokens_before = sum(b.token_count for b in blocks)

    if config.strategy == CompressionStrategy.SLIDING_WINDOW:
        kept = _sliding_window(blocks, config.keep_recent or config.preserve_recent, config.preserve_recent)
    elif config.strategy == CompressionStrategy.WEIGHTED_PRUNE:
        kept = _weighted_prune(blocks, config.target_tokens, config.preserve_recent)
    else:  # HYBRID: prune first, then window if still over budget
        after_prune = _weighted_prune(blocks, config.target_tokens, config.preserve_recent)
        if sum(b.token_count for b in after_prune) > config.target_tokens:
            kept = _sliding_window(after_prune, config.keep_recent or config.preserve_recent, config.preserve_recent)
        else:
            kept = after_prune

    tokens_after = sum(b.token_count for b in kept)
    return SurgeonResult(
        blocks=kept,
        tokens_before=tokens_before,
        tokens_after=tokens_after,
        tokens_saved=tokens_before - tokens_after,
        strategy_applied=config.strategy.value,
    )


def _sliding_window(blocks: list[ContextBlock], keep_recent: int, preserve_recent: int) -> list[ContextBlock]:
    if keep_recent == 0:
        return list(blocks)
    always_keep = max(preserve_recent, keep_recent)
    if len(blocks) <= always_keep:
        return list(blocks)
    pinned = [b for b in blocks if b.weight >= ContextWeight.HIGH]
    pinned_ids = {b.id for b in pinned}
    recent_start = max(0, len(blocks) - keep_recent)
    result = list(pinned)
    for b in blocks[recent_start:]:
        if b.id not in pinned_ids:
            result.append(b)
    return result


def _weighted_prune(blocks: list[ContextBlock], target_tokens: int, preserve_recent: int) -> list[ContextBlock]:
    total = sum(b.token_count for b in blocks)
    if total <= target_tokens:
        return list(blocks)

    preserve_ids = {b.id for b in blocks[-preserve_recent:]} if preserve_recent > 0 else set()

    removable = sorted(
        [b for b in blocks if b.id not in preserve_ids and b.weight < ContextWeight.HIGH],
        key=lambda b: (int(b.weight), b.created_at),
    )

    to_remove: set = set()
    running = total
    for b in removable:
        if running <= target_tokens:
            break
        running -= b.token_count
        to_remove.add(b.id)

    return [b for b in blocks if b.id not in to_remove]


def score_relevance(block: ContextBlock, task: str) -> float:
    """Keyword overlap between block content and task string. Returns 0.0-1.0."""
    task_words = set(task.split())
    if not task_words:
        return 0.0
    matches = sum(1 for w in block.content.split() if w in task_words)
    return min(matches / len(task_words), 1.0)


def detect_contradiction(a: ContextBlock, b: ContextBlock) -> bool:
    """Heuristic: True if A asserts X and B asserts 'not X'.

    Single-direction by design: checks words from A against negations in B only.
    This avoids false positives where A contains incidental negations (e.g. "do not press").
    """
    a_lower = a.content.lower()
    b_lower = b.content.lower()
    for word in a_lower.split():
        if len(word) > 4 and f"not {word}" in b_lower:
            return True
    return False
